packages/devtools/src/schema/resolution.ts

```diff
@@ -0,0 +1,340 @@
+      : Promise.all(sourceDefinition.implements.map(getDependencyID)).then((ids) => {
+          return loadAllModelInterfaces(ceramic, ids)
+        })
+
+    // Resolve named dependencies in relations to their stream ID
+    const relationsPromise = promiseMap(sourceDefinition.relations ?? {}, async (relation) => {
+      return relation.type === 'document' && relation.model !== null
+        ? { ...relation, model: await getDependencyID(relation.model) }
+        : relation
+    })
+
+    // Resolve named dependencies in views to their stream ID when possible, or move the view to the composite
+    const compositeViews: ModelViewsDefinitionV2 = {}
+    const viewsPromises: Record<string, Promise<ModelViewDefinitionV2>> = {}
+    for (const [name, view] of Object.entries(sourceDefinition.views ?? {})) {
+      if (
+        (view.type === 'relationCountFrom' ||
+          view.type === 'relationFrom' ||
+          view.type === 'relationDocument') &&
+        view.model !== null
+      ) {
+        const existing = executing[view.model]
+        if (existing == null) {
+          compositeViews[name] = view
+        } else {
+          viewsPromises[name] = getDependencyID(view.model).then((model) => {
+            return { ...view, model }
+          })
+        }
+      } else {
+        viewsPromises[name] = Promise.resolve(view)
+      }
+    }
+
+    const [implementIDs, relations, views] = await Promise.all([
+      implementsPromise,
+      relationsPromise,
+      promiseMap(viewsPromises, (viewPromise) => viewPromise),
+    ])
+    // Always convert to a v2 definition
+    const newDefinition: ModelDefinitionV2 = {
+      version: '2.0',
+      name: sourceDefinition.name,
+      description: sourceDefinition.description,
+      accountRelation: sourceDefinition.accountRelation,
+      interface: isV1 ? false : sourceDefinition.interface,
+      implements: implementIDs,
+      schema: sourceDefinition.schema,
+      relations,
+      views,
+    }
+
+    const model = await Model.create(ceramic, newDefinition)
+    assertSupportedWriteModelController(model, ceramic)
+    const id = model.id.toString()
+
+    return {
+      id,
+      name: modelName,
+      model,
+      commitsPromise: loadCommits(ceramic, id),
+      indices: definition.indices ?? [],
+      views: compositeViews,
+    }
+  }
+}
+
+function executeLoadFactory(
+  ceramic: CeramicApi,
+  modelName: string,
+  definition: AbstractLoadModelDefinition,
+) {
+  return async function executeLoad(): Promise<ResolvedModel> {
+    const id = definition.id
+    const model = await Model.load(ceramic, id)
+    assertSupportedWriteModelController(model, ceramic)
+    assertValidModelInterfaceType(model.content, definition.interface)
+
+    return {
+      id,
+      name: modelName,
+      model,
+      commitsPromise: loadCommits(ceramic, id),
+      indices: definition.indices ?? [],
+      views: definition.views,
+    }
+  }
+}
+
+function assertNoCircularDependency(
```

Comment:
just one question, do we have to do this every time we load the model, or only when creating it?  Just don't want to have too much overhead if we have to test this every time

Reply:
It's only needed on creation because we don't support creating models with circular relations anyways, so here the logic is about making sure the composite the user attempts to define can actually be created.